module2/homework/grayscaleImages/compression.cpp

```diff
@@ -0,0 +1,53 @@
+#include "compression.hpp"
+
+#include <algorithm>
+#include <cstdint>
+#include <iterator>
+#include <utility>
+
+compressedImage compressGrayscale(decompressedImage& input) {
+    compressedImage result;
+    result.reserve(width * height);
+
+    std::for_each(begin(input), end(input), [&result](decompressedLine& line){
+        compressedImage compressedLine = compressSingleLine(line);
+        std::transform(begin(compressedLine), end(compressedLine), std::back_inserter(result), [](const auto& pair){
+            return std::make_pair(pair.first, pair.second);
+        });
+    });
+
+    result.shrink_to_fit();
+    return result;
+}
+
+compressedImage compressSingleLine(decompressedLine& line) {
+    compressedImage buffer;
+    buffer.reserve(width);
+    std::pair<uint8_t, uint8_t> current = {line.front(), 0};
+
+    std::for_each(begin(line), end(line), [&buffer, &current](const uint8_t pixel){
+        if (pixel == current.first) {
+            current.second++;
+        }
+        else {
+            buffer.push_back(current);
+            current = {pixel, 1};
+        }
+    });
+
+    buffer.push_back(current);
```

Comment:
You should try to refactorize it,to have buffer.push_back(current) once.

Reply:
+1